src/sysdolphin/baselib/jobj.c

```diff
@@ -23,3 +27,249 @@ void HSD_JObjCheckDepend(HSD_JObj* jobj)
+        HSD_JObj* aobj_jobj = (HSD_JObj*) jobj->aobj->hsd_obj;
+        HSD_JObjSetupMatrix((HSD_JObj*) jobj->aobj->hsd_obj);
+        PSMTXMUltiVec(aobj_jobj->mtx, &jobj->translate, &vec);
+        jobj->mtx[0][3] = vec.x;
+        jobj->mtx[1][3] = vec.y;
+        jobj->mtx[2][3] = vec.z;
+    }
+}
+
+void HSD_JObjRemoveAnimByFlags(HSD_JObj* jobj, u32 flags)
+{
+    BOOL has_dobj;
+    if (jobj != NULL) {
+        if (flags & 1) {
+            HSD_AObjRemove(jobj->aobj);
+            jobj->aobj = NULL;
+        }
+        if (jobj->flags & 0x4020) {
+            has_dobj = FALSE;
+        } else {
+            has_dobj = TRUE;
+        }
+        if (has_dobj) {
+            HSD_DObjRemoveAnimAllByFlags(jobj->u.dobj, flags);
+        }
+        HSD_RObjRemoveAnimAllByFlags(jobj->robj, flags);
+    }
+}
+
+void HSD_JObjRemoveAnimAllByFlags(HSD_JObj* jobj, u32 arg1)
+{
+    if (jobj != NULL) {
+        HSD_JObjRemoveAnimByFlags(jobj, arg1);
+        if (!(jobj->flags & 0x1000)) {
+            HSD_JObj* child_jobj = jobj->child;
+            while (child_jobj != NULL) {
+                HSD_JObjRemoveAnimAllByFlags(child_jobj, arg1);
+                child_jobj = child_jobj->next;
+            }
+        }
+    }
+}
+
+void HSD_JObjRemoveAnim(HSD_JObj* jobj)
+{
+    HSD_JObjRemoveAnimByFlags(jobj, 0x7FF);
+}
+
+void HSD_JObjRemoveAnimAll(HSD_JObj* jobj)
+{
+    HSD_JObjRemoveAnimAllByFlags(jobj, 0x7FF);
+}
+
+void HSD_JObjReqAnimByFlags(HSD_JObj* jobj, u32 flags, f32 frame)
+{
+    BOOL has_dobj;
+    if (jobj != NULL) {
+        if (flags & 1) {
+            HSD_AObjReqAnim(jobj->aobj, frame);
+        }
+        if (jobj->flags & 0x4020) {
+            has_dobj = FALSE;
+        } else {
+            has_dobj = TRUE;
+        }
+        if (has_dobj) {
+            HSD_DObjReqAnimAllByFlags(jobj->u.dobj, frame, flags);
+        }
+        HSD_RObjReqAnimAllByFlags(jobj->robj, flags, frame);
+    }
+}
+
+void HSD_JObjReqAnimAllByFlags(HSD_JObj* jobj, u32 flags, f32 frame)
+{
+    if (jobj != NULL) {
+        HSD_JObjReqAnimByFlags(jobj, flags, frame);
+        if (!(jobj->flags & 0x1000)) {
+            HSD_JObj* child = jobj->child;
+            while (child != NULL) {
+                HSD_JObjReqAnimAllByFlags(child, flags, frame);
+                child = child->next;
+            }
+        }
+    }
+}
+
+void HSD_JObjReqAnimAll(HSD_JObj* jobj, f32 frame)
+{
+    HSD_JObjReqAnimAllByFlags(jobj, 0x7FF, frame);
+}
+
+void HSD_JObjReqAnim(HSD_JObj* jobj, f32 frame)
+{
+    HSD_JObjReqAnimByFlags(jobj, 0x7FF, frame);
+}
+
+void JObjSortAnim(HSD_AObj* aobj)
+{
+    HSD_FObj* fobj;
+    HSD_FObj** fobj_p;
+
+    if (aobj == NULL || aobj->fobj == NULL) {
+        return;
+    }
+    fobj_p = &aobj->fobj;
+    while ((fobj = *fobj_p) != NULL) {
+        if (fobj->obj_type == 0xC) {
+            *fobj_p = fobj->next;
+            fobj->next = aobj->fobj;
+            aobj->fobj = fobj;
+            break;
+        }
+        fobj_p = &fobj->next;
```

Comment:
This reads better to me:
```suggestion
    HSD_FObj* fobj;
    HSD_FObj** cur;

    if (aobj == NULL || aobj->fobj == NULL) {
        return;
    }
    cur = &aobj->fobj;
    while ((fobj = *cur) != NULL) {
        if (fobj->obj_type == 0xC) {
            *cur = fobj->next;
            fobj->next = aobj->fobj;
            aobj->fobj = fobj;
            break;
        }
        cur = &fobj->next;
```

Reply:
I'd rather keep `fobj_p` or similar to indicate that this is the same as `fobj`. This is a weird pattern where the pointer and pointer-pointer are both needed to match.

Reply:
Are you ok with `fobj_ptr`?